sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -171,13 +172,23 @@ case class FilterExec(condition: Expression, child: SparkPlan)
     // This is very perf sensitive.
     // TODO: revisit this. We can consider reordering predicates as well.
     val generatedIsNotNullChecks = new Array[Boolean](notNullPreds.length)
+
+    val extraIsNotNullReferences = mutable.Set[Attribute]()
+
+    def outputContainsNotNull(ref: Attribute): Boolean = {
```

Comment:
This doesn't mean the attribute is not nullable. We should check `notNullAttributes.contains(ref.exprId)`

Reply:
Ah, my mistake.